apps/ejabberd/src/cyrsasl.erl

```diff
@@ -36,15 +36,12 @@
 
 -include("ejabberd.hrl").
 
--record(sasl_mechanism, {mechanism :: mechanism(),
+-record(sasl_mechanism, {mechanism :: mechanism() | atom(), % atom() used in ets:select
                          module :: sasl_module(),
-                         password_type :: plain | digest | scram
+                         password_type :: plain | digest | scram | atom() % atom() used in ets:select
                         }).
--type sasl_module() :: cyrsasl_anonymous
-                     | cyrsasl_digest
-                     | cyrsasl_plain.
+-type sasl_module() :: atom().
```

Comment:
I agree with @fenek. This is basically an Erlang equivalent of OO-style subtyping, i.e. in such a case we're striving for possibly infinite extensibility (with regard to number of implementations - the interface of each implementation is frozen). We must have a corresponding behaviour describing the set of callbacks expected from a `sasl_module()`, though. It should be `sasl_module() :: module()`.